Improve Batchee with Performance Boosts and Error Handling

Yusufibin opened the pull request:
### **Main Changes**

- **Code Quality**: Added check for empty items and tests for edge cases (empty files, invalid inputs).

**### Tests**

- All existing tests pass.

- New tests added for edge cases.

- Verified backward compatibility.

Yusufibin committed:
update

danielfromearth commented:
Hi @Yusufibin, thanks for your interest in this repository and for sharing these proposed code changes! 

In particular, I believe the following three changes will be improvements to the robustness of the `batchee` package:

1. in [batchee/harmony/util.py](https://github.com/nasa/batchee/pull/197/files#diff-9846a162c1963d23679df7876c5757fd3f078f9a7945857545041b902728b013)
```python
    if not input_items:
        return []
```

2. in [‎tests/test_filename_grouping.py‎](https://github.com/nasa/batchee/pull/197/files#diff-448c165a2ea73d8f0e30217a1f67fe6ecb19dceaf6ed70fc84edf3e925a60f57)
```python
def test_empty_list():
    results = get_batch_indices([])
    assert results == []
```

3.  in [‎tests/test_filename_grouping.py‎](https://github.com/nasa/batchee/pull/197/files#diff-448c165a2ea73d8f0e30217a1f67fe6ecb19dceaf6ed70fc84edf3e925a60f57)
```python
def test_invalid_filenames():
    invalid_filenames = ["invalid.nc", "TEMPO_NO2_L2_V03_20240731.nc", "random_file.txt"]
    results = get_batch_indices(invalid_filenames)
    assert results == []
```

I believe the other changes will add unnecessary complexity.  If you are able to remove the other changes, we should be able to then merge these three above changes into the develop branch.

Yusufibin committed:
update

Yusufibin committed:
update

Yusufibin commented:
Hello @danielfromearth ,
Thank you for your feedback on the pull request. I reviewed the code and removed the other changes to avoid unnecessary complexity, keeping only the three approved improvements to enhance the robustness of the **batchee** package:

1. In **batchee/harmony/util.py**: Added `if not input_items: return []` in the `_get_output_bounding_box` function to handle empty lists.
2. In **tests/test_filename_grouping.py**: Added the `test_empty_list` test to verify behavior with an empty list.
3. In **tests/test_filename_grouping.py**: Added the `test_invalid_filenames` test to verify behavior with invalid filenames.

All other changes (URL validation, performance optimizations, output formats, high-volume testing, etc.) have been removed. The tests pass successfully, confirming that the code remains functional.

The pull request is now ready for potential merging into the **develop** branch.

danielfromearth commented on batchee/tempo_filename_parser.py:
```diff
@@ -87,8 +87,6 @@ def get_batch_indices(filenames: list, logger: logging.Logger = default_logger)
     list[int]
         batch index for each filename in the original list, e.g. [0, 0, 0, 1, 1, 1, ...]
     """
-    logger.info(f"get_batch_indices() starting --- with {len(filenames)} filenames")
```
I think this `logger.info()` line should be kept in, as it is helpful for troubleshooting.

danielfromearth reviewed (changes requested):
@Yusufibin, this looks good now. Thank you for making the updates!

There's one line here that I think should be reverted to how it was before, but otherwise it will be ready to go (pending new CHANGELOG entry and tests passing).

Yusufibin committed:
Update tempo_filename_parser.py

Yusufibin committed:
Update tempo_filename_parser.py

Yusufibin commented:
@danielfromearth Done

danielfromearth commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

danielfromearth commented on batchee/tempo_filename_parser.py:
```diff
@@ -87,9 +87,9 @@ def get_batch_indices(filenames: list, logger: logging.Logger = default_logger)
+
     logger.info(f"get_batch_indices() starting --- with {len(filenames)} filenames")
 
-    # Make a new list with days and scans, e.g. [('20130701', 'S009'), ('20130701', 'S009'), ...]
```
@Yusufibin, can you put this comment back in, it seems to me like a helpful description?

Yusufibin replied:
@danielfromearth Done

Yusufibin committed:
Add comment for day_and_scans list clarification

Added comment to clarify the purpose of the day_and_scans list.

Yusufibin committed:
Update tempo_filename_parser.py

danielfromearth commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
